File: examples/BoxNPSAT/Box3DNPSAT.py
import matplotlib.pyplot as plt
import random
import math
import numpy

# Domain outline
Dx = [0, 5000, 5000, 0, 0]
Dy = [0, 0, 5000, 5000, 0]
plt.plot(Dx, Dy, label='Outline')

# river segments
r1x = [1178, 2427, 1912, 2244, 3032]
r1y = [5000, 3568, 2877, 1808, 1200]
plt.plot(r1x, r1y, label='River 1')

r2x = [5000, 4052, 4329, 3200]
r2y = [3442, 2630, 1463, 1032]
plt.plot(r2x, r2y, label='River 2')

# Constant head boundary
Cdelta = 200
Cp = [3000, 1000]

cx = [Cp[0] - Cdelta, Cp[0] + Cdelta, Cp[0] + Cdelta, Cp[0] - Cdelta, Cp[0] - Cdelta]
cy = [Cp[1] - Cdelta, Cp[1] - Cdelta, Cp[1] + Cdelta, Cp[1] + Cdelta, Cp[1] - Cdelta]
plt.plot(cx, cy, label='Constant head')

# Recharge
N = 0.00018
Qrch = 5000*5000*N
print("Q Recharge = " + str(Qrch))

# Total stream recharge is 20% of the areal recharge
Qstream = Qrch * 0.2
print("Q Stream = " + str(Qstream))

Riv_len1 = 0.0
R_seg_len1 = []
for ii in range(0, len(r1x)-1):
    R_seg_len1.append(math.sqrt((r1x[ii+1] - r1x[ii])**2 + (r1y[ii+1] - r1y[ii])**2))
    Riv_len1 = Riv_len1 + R_seg_len1[-1]
print(Riv_len1)

Riv_len2 = 0.0
R_seg_len2 = []
for ii in range(0, len(r2x)-1):
    R_seg_len2.append(math.sqrt((r2x[ii+1] - r2x[ii])**2 + (r2y[ii+1] - r2y[ii])**2))
    Riv_len2 = Riv_len2 + R_seg_len2[-1]
print(Riv_len2)

# distribute according to river length
# Qstrm1 = Qstream * (Riv_len1 / (Riv_len1 + Riv_len2))
# Qstrm2 = Qstream * (Riv_len2 / (Riv_len1 + Riv_len2))\

# arbitrarily set 70% of the river to come from the 1st river
Qstrm1 = Qstream * 0.7
Qstrm2 = Qstream * 0.3

print("Qriver 1 = " + str(Qstrm1) + ' Qriver 2 = ' + str(Qstrm2))

Total_pumping = Qstrm1 + Qstrm2 + Qrch
print("Total pumping required " + str(Total_pumping))

# Generate as many wells as needed to equalize the pumping
Q_so_far = 0

wx = []
wy = []
Q = []
SL = []
Top = []
for ii in range(0, 200):
    x = random.uniform(300.0, 4700.0)
    y = random.uniform(300.0, 4700.0)

    if x > cx[0] and x < cx[1] and y > cy[0] and y < cy[2]:
        continue

    q = numpy.random.normal(150, 50)
    q = numpy.clip(q, 10, 250)

    sl = random.uniform(0.5, 1) * q
    t = random.uniform(0.1, 0.7)* sl

    well_exist = False
    if ii == 0:
        wx.append(x)
        wy.append(y)
    else:
        for iw in range(0, len(wx)):
            wdst = math.sqrt((x - wx[iw])**2 + (y - wy[iw])**2)
            # print(wdst)
            if wdst < 200:
                well_exist = True
                break

    if not well_exist:
        wx.append(x)
        wy.append(y)
        Q.append(q)
        SL.append(sl)
        Top.append(t)
        # print('Q = ' + str(q) + ', L = ' + str(sl) + ', T = ' + str(t) + ', B = ' +  str(sl + t))
        Q_so_far = Q_so_far + q

    if Q_so_far > Total_pumping:
        break

print(Q_so_far)
plt.plot(wx,wy, 'ro', label='Wells')

plt.show()

# -------------------PRINT FILES--------------------
# WELL file
f = open('box01_wells.npsat', 'w+')
f.write('%d\r\n' % len(wx))
Gelev = 50

for ii in range(0, len(wx)-1):
    f.write('%.2f\t%.2f\t%.2f\t%.2f\t%.2f\r\n' % (wx[ii], wy[ii], Gelev - Top[ii],  Gelev - Top[ii] -SL[ii], -Q[ii]))
f.close()

# RIVER file
f = open('box01_rivers.npsat', 'w+')
N_riv_seg = len(r1x) - 1 + len(r2x) - 1
f.write('%d\r\n' % N_riv_seg)

for ii in range(0, len(r1x)-1):
    Q_riv_seg = Qstrm1*(R_seg_len1[ii]/Riv_len1)/(Riv_len1*50.0)
    print(Qstrm1*(R_seg_len1[ii]/Riv_len1)/(Riv_len1*50.0))
    f.write('%.2f\t%.2f\t%.2f\t%.2f\t%.6f\t%.2f\r\n' %(r1x[ii], r1y[ii], r1x[ii+1], r1y[ii+1], Q_riv_seg, 50))

for ii in range(0, len(r2x)-1):
    Q_riv_seg = Qstrm2*(R_seg_len2[ii]/Riv_len2)/(Riv_len2*30.0)
    f.write('%.2f\t%.2f\t%.2f\t%.2f\t%.6f\t%.2f\r\n' %(r2x[ii], r2y[ii], r2x[ii+1], r2y[ii+1], Q_riv_seg, 30))

f.close()

File: examples/BoxNPSAT/BOX3DNPSAT.m
%% Aquifer outline
domain = [0 0; 5000 0; 5000 5000; 0 5000];
%% River segments
r1 = [1178 5000; 2427 3568; 1912 2877; 2244 1808; 3032 1200];
r2 = [5000 3442; 4052 2630; 4329 1463; 3200 1032];
%% Boundary conditions
bc_delta = 200;
bc_center = [3000 1000];
bc_poly = [...
           bc_center(1) -  bc_delta bc_center(2) -  bc_delta; ... 
           bc_center(1) +  bc_delta bc_center(2) -  bc_delta; ...
           bc_center(1) +  bc_delta bc_center(2) +  bc_delta; ...
           bc_center(1) -  bc_delta bc_center(2) +  bc_delta; ...
];
%% Recharge 
Rch = 0.00018; % m/day
Qrch = (5000*5000 - polyarea(bc_poly(:,1), bc_poly(:,2)))*Rch;
%% Stream recharge
% Let's suppose that there is an additional 20% of diffuse recharge coming
% from rivers
Qstream = Qrch*0.2;
r1_seg_len = sqrt(diff(r1(:,1)).^2 + diff(r1(:,2)).^2);
r2_seg_len = sqrt(diff(r2(:,1)).^2 + diff(r2(:,2)).^2);
% let's also assume that 70% of the total stream recharge comes from the
% river 1
Qstrm1 = Qstream * 0.7;
Qstrm2 = Qstream * 0.3;
%% Generate pumping wells
Tot_pump = Qrch + Qstream;
Tot_Qw = 0;
wells = [];
Rseg = [[r1(1:size(r1,1)-1,:) r1(2:size(r1,1),:)]; ...
        [r2(1:size(r2,1)-1,:) r2(2:size(r2,1),:)]];
while Tot_Qw < Tot_pump
    while true
        xw = 400 + (4600 - 400)*rand;
        yw = 400 + (4600 - 400)*rand;
        if ~isempty(wells) 
            mindst = min(sqrt((wells(:,1) - xw).^2 + (wells(:,2) - yw).^2));
            if mindst < 400
                continue;
            end
        end

        dst = sqrt((xw-bc_center(1)).^2 + (yw-bc_center(2)).^2);
        if dst < bc_delta+400
            continue;
        end
        dst = Dist_Point_LineSegment(xw,yw,Rseg);
        if min(dst) < 400
            continue;
        else
            break;
        end
        
    end
    qw = 100 + (500 - 100)*rand;
    slw = 10 + (100-10)*rand;
    r = (250 - slw)*rand;
    bw = -270 + r;
    tw = bw + slw;
    wells = [wells;xw yw tw bw -qw];
    Tot_Qw = Tot_Qw + qw;
end
wells(:,5) = wells(:,5).*(Tot_pump/sum(wells(:,5)));
%% plot 
clf 
hold on
plot(domain([1:4 1],1), domain([1:4 1],2), 'linewidth',2, 'DisplayName', 'No flow')
plot([r1(:,1);nan;r2(:,1)], [r1(:,2);nan;r2(:,2)], 'linewidth',2, 'DisplayName', 'Rivers') 
plot(bc_poly([1:4 1],1), bc_poly([1:4 1],2), 'linewidth',2, 'DisplayName', 'Constant head')
plot(wells(:,1), wells(:,2),'.', 'DisplayName', 'Wells')
legend('Location','northoutside','Orientation','horizontal')
axis equal
axis off
%% Write files
% wells
fid = fopen('box01_wells.npsat','w');
fprintf(fid, '%d\n', size(wells,1));
fprintf(fid, '%f %f %f %f -%f\n', wells');
fclose(fid);
%% Boundary Conditions
fid = fopen('box01_bc.npsat','w');
fprintf(fid, '1\n');
fprintf(fid, 'TOP %d %.2f\n', [size(bc_poly,1) 30]);
fprintf(fid, '%.2f %.2f\n', bc_poly');
fclose(fid);
%% Rivers
fid = fopen('box01_rivers.npsat', 'w');
fprintf(fid, '%d\n', length(r1_seg_len) + length(r2_seg_len));
perc = r1_seg_len/sum(r1_seg_len).*(0.3 + 0.4*rand(length(r1_seg_len),1));
perc = perc/sum(perc);
for ii = 1:length(r1_seg_len)
    fprintf(fid, '%d %.5f %.2f\n', [2 Qstrm1*perc(ii)/(50*r1_seg_len(ii)) 50]);
    fprintf(fid, '%.2f %.2f\n', r1(ii,:));
    fprintf(fid, '%.2f %.2f\n', r1(ii+1,:));
end
perc = r2_seg_len/sum(r2_seg_len).*(0.3 + 0.4*rand(length(r2_seg_len),1));
perc = perc/sum(perc);
for ii = 1:length(r2_seg_len)
    fprintf(fid, '%d %.5f %.2f\n', [2 Qstrm2*perc(ii)/(50*r2_seg_len(ii)) 50]);
    fprintf(fid, '%.2f %.2f\n', r2(ii,:));
    fprintf(fid, '%.2f %.2f\n', r2(ii+1,:));
end
fclose(fid);
%% Write initial particle locations
xx = 100:100:4900;
yy = 100:100:4900;
zz = -250:50:0;
[Xpart, Ypart, Zpart] = meshgrid (xx, yy, zz);
Nx = length(100:100:4900);
Ny = length(100:100:4900);
Nz = length(-250:50:0);
Xpart = reshape(Xpart,Nx*Ny*Nz,1);
Ypart = reshape(Ypart,Nx*Ny*Nz,1);
Zpart = reshape(Zpart,Nx*Ny*Nz,1);
Eid = nan(length(Xpart),1);
Sid = nan(length(Xpart),1);
% assign different eintity id for each level
for ii = 1:Nz
   idz = find(Zpart == zz(ii));
   Eid(idz) = ii;
   Sid(idz) = [1:length(idz)]';
end
%% write particle locations
fid = fopen('layer_part.npsat','w');
fprintf(fid, '%d\n', length(Eid));
fprintf(fid, '%d %d %.1f %.1f %.1f\n', [Eid Sid Xpart Ypart Zpart]');
fclose(fid);

    


